File: analyze_protein_trajectory/HP36/traj_A/average_ent_vs_residue/average_ent_vs_residue.f
c
c     calculate the average of ent_vs_residue
c
c       input  : list_files_ent_res
c       output : average_ent_vs_residue.dat
c

      program statistics_ent_vs_residue

      implicit none

      integer maxres
      parameter (maxres=1000)

      integer nres, ndata
      real*8  ave_ent_tot(maxres)
      real*8  ave_ent_LJ(maxres)
      real*8  ave_ent_cou(maxres)

      real*8  ave2_ent_tot(maxres), sig_ent_tot

      real*8  ent_res, ent_res_LJ, ent_res_cou
      integer ires, ires_dum

      real*8  ave_ent_all

      character*200 filename

c
c     initialization
c

      ndata = 0

      do ires=1,maxres

        ave_ent_tot(ires) = 0.d0
        ave_ent_LJ(ires)  = 0.d0
        ave_ent_cou(ires) = 0.d0

        ave2_ent_tot(ires) = 0.d0

      enddo

      ave_ent_all = 0.d0

c
c     read data and caluclate the statistics
c

      open(10,file='list_files_ent_res',status='old')

  100 read(10,'(A)',end=190,err=90) filename
      open(20,file=filename,status='old',err=90)

        ndata = ndata + 1

        read(20,*,err=90) nres

        if (nres.gt.maxres)  then
          print *,'error: number of residues exceeds maxres'
          goto 99
        endif

        do ires=1,nres

          read(20,*,err=90) ires_dum,ent_res,ent_res_LJ,ent_res_cou

          if (ires_dum.ne.ires)  then
            print *,'error:  inconsistency in the residue number' 
            goto 99
          endif

          ave_ent_tot(ires) = ave_ent_tot(ires) + ent_res
          ave_ent_LJ(ires)  = ave_ent_LJ(ires)  + ent_res_LJ
          ave_ent_cou(ires) = ave_ent_cou(ires) + ent_res_cou

          ave2_ent_tot(ires) = ave2_ent_tot(ires) + ent_res*ent_res

          ave_ent_all = ave_ent_all + ent_res

        enddo

        close(20,err=90)

      goto 100

  190 continue

      close(10,err=90)

c     normalization

      do ires=1,nres

        ave_ent_tot(ires) = ave_ent_tot(ires) / ndata
        ave_ent_LJ(ires)  = ave_ent_LJ(ires)  / ndata
        ave_ent_cou(ires) = ave_ent_cou(ires) / ndata

        ave2_ent_tot(ires) = ave2_ent_tot(ires) / ndata

      enddo

      ave_ent_all = ave_ent_all / ndata

c
c     save average_ent_vs_residue.dat
c

      open(30,file='average_ent_vs_residue.dat',err=90)

      do ires=1,nres

        sig_ent_tot = dsqrt(ave2_ent_tot(ires)-ave_ent_tot(ires)**2)

        write(30,990) ires, ave_ent_tot(ires),
     .                      ave_ent_LJ(ires),
     .                      ave_ent_cou(ires),
     .                      sig_ent_tot

      enddo

      close(30,err=90)

c
c     print out summary
c

c     on screen

      write(6,*)
      write(6,*) 'total number of data used = ', ndata
      write(6,*)
      write(6,*) 'ave_ent_all         : ', ave_ent_all
      write(6,*)

c     on file

      open(30,file='summary_average_ent_vs_residue.dat',err=90)

      write(30,*)
      write(30,*) 'total number of data used = ', ndata
      write(30,*)
      write(30,*) 'ave_ent_all         : ', ave_ent_all
      write(30,*)

      close(30,err=90)

c
c     format
c

  990 format(i5,1x,f16.8,1x,f16.8,1x,f16.8,1x,f16.8)

c..........
   99 stop
   90 print *,'MAIN:  I/O error'
      goto 99

      end

File: analyze_protein_trajectory/py_tools_monomer/average_ent_vs_residue/average_ent_vs_residue.py
import sys
import glob
import numpy as np

"""
calculate the average of ent_vs_residue

input  : path of ent_res files in thm decomposition directory
         "../../traj_A/3D-RISM/thm_decomposition/thma_data/"

output : average_ent_vs_residue.dat

"""

dat_dir = sys.argv[1]

list_files = sorted(glob.glob('{}/*/ent_res.dat'.format(dat_dir)))
ndata = len(list_files)

# read data and caluclate the statistics
data = []
ent_all = []
for k, file in enumerate(list_files):
    ent_sum = 0.

    with open(file, 'r') as f:
        lines = f.readlines()[1:]  # skip first line (num. of residues)
        for line in lines:
            line = line.replace('\n','')
            ent_res = float(line[6:23])
            ent_sum += ent_res
            data.append(line)

    ent_all.append(ent_sum)

    if k == 0:
        n_res = len(data)
        print('number of residues    =', n_res)


print('total number of data  =', ndata)
print('average_ent_all       =', np.mean(ent_all))

f_out = open('summary_average_ent_vs_residue.dat', 'w')
print('total number of data  =', ndata, file=f_out)
print('', file=f_out)
print('average_ent_all       =', np.mean(ent_all), file=f_out)
print('', file=f_out)
f_out.close()


# calculate the average of ent_vs_residue
f_out = open('average_ent_vs_residue.dat', 'w')

for i in range(1, n_res+1):
    # select data by residue number
    res_x = [x for x in data if int(x[:6]) == i]

    l_ent_res = []
    l_ent_LJ  = []
    l_ent_cou = []
    l_ent_sig = []

    for line in res_x:
        ent_res = float(line[6:23])
        ent_res_LJ = float(line[23:40])
        ent_res_cou = float(line[40:57])

        l_ent_res.append(ent_res)
        l_ent_LJ.append(ent_res_LJ)
        l_ent_cou.append(ent_res_cou)
        l_ent_sig.append(ent_res*ent_res)

    ave_ent_tot = np.mean(l_ent_res)
    ave_ent_LJ = np.mean(l_ent_LJ)
    ave_ent_cou = np.mean(l_ent_cou)

    sig_ent_tot = np.sqrt(np.mean(l_ent_sig) - ave_ent_tot**2)

    print('{:>6d} {:16.8f} {:16.8f} {:16.8f} {:16.8f}'.format(i,
                            ave_ent_tot, ave_ent_LJ, ave_ent_cou, sig_ent_tot),
                            file=f_out)

f_out.close()

# =============================================================================
